src/services/organizeImports.ts

```diff
@@ -577,30 +590,71 @@ function getModuleSpecifierExpression(declaration: AnyImportOrRequireStatement):
 }
 
 /** @internal */
-export function importsAreSorted(imports: readonly AnyImportOrRequireStatement[]): imports is SortedReadonlyArray<AnyImportOrRequireStatement> {
-    return arrayIsSorted(imports, compareImportsOrRequireStatements);
+export function detectSorting(sourceFile: SourceFile): SortKind {
+    return detectSortingWorker(
+        groupImportsByNewlineContiguous(sourceFile, sourceFile.statements.filter(isImportDeclaration)));
+}
+
+function detectSortingWorker(importGroups: ImportDeclaration[][]): SortKind {
+    let sortState = SortKind.Both;
+    for (const importGroup of importGroups) {
+        // Check module specifiers
+        if (importGroup.length > 1) {
+            sortState &= detectSortCaseSensitivity(importGroup, /*useEslintOrdering*/ true, i => tryCast(i.moduleSpecifier, isStringLiteral)?.text ?? "");
+            if (!sortState) {
+                return sortState;
+            }
+        }
+
+        // Check import specifiers
+        const declarationWithNamedImports = find(
+            importGroup,
+            i => tryCast(i.importClause?.namedBindings, isNamedImports)?.elements.length! > 1);
+        if (declarationWithNamedImports) {
+            sortState &= detectImportSpecifierSorting((declarationWithNamedImports.importClause!.namedBindings as NamedImports).elements);
+            if (!sortState) {
+                return sortState;
+            }
+        }
+
+        // Quit as soon as we've disambiguated. There's a chance that something later will disagree with what we've
+        // found so far, but this function is only intended to infer a preference, not validate the whole file for
+        // consistent and correct sorting.
+        if (sortState !== SortKind.Both) {
+            return sortState;
+        }
+    }
+    return sortState;
 }
 
 /** @internal */
-export function importSpecifiersAreSorted(imports: readonly ImportSpecifier[]): imports is SortedReadonlyArray<ImportSpecifier> {
-    return arrayIsSorted(imports, compareImportOrExportSpecifiers);
+export function detectImportDeclarationSorting(imports: readonly AnyImportOrRequireStatement[]): SortKind {
+    return detectSortCaseSensitivity(imports, /*useEslintOrdering*/ true, s => getExternalModuleName(getModuleSpecifierExpression(s)) || "");
 }
 
 /** @internal */
-export function getImportDeclarationInsertionIndex(sortedImports: SortedReadonlyArray<AnyImportOrRequireStatement>, newImport: AnyImportOrRequireStatement) {
-    const index = binarySearch(sortedImports, newImport, identity, compareImportsOrRequireStatements);
+export const detectImportSpecifierSorting = memoizeWeak((specifiers: readonly ImportSpecifier[]): SortKind => {

```

Comment:
Is there a particular reason why this one is cached but others aren't? Would everything benefit from being cached like this?

Reply:
As it stands now, this is the only one that benefits from being cached. It gets called several times from auto-imports in ways that are cumbersome to cache locally.